### ISSUE TITLE
Question regarding acquiring the provider name from an EdDSA EVP_MD_CTX using 3.0 beta 2

### ISSUE THREAD

**Issue by dskern-github** [openssl/openssl#16233]
**TL; DR:** 

1. Thanks for adding `EVP_PKEY_get0_provider` and `EVP_PKEY_CTX_get0_provider` to beta 2. I've exhaustively tested the latter with RSA encryption and decryption and it works great. 
2. `EVP_MD_CTX_get0_md` -> `EVP_MD_get0_provider` works for RSA and ECDSA signing and verification but fails with EdDSA, presumably because PureEdDSA does not include a pre-hash step.  
3. `EVP_MD_CTX_get_pkey_ctx` -> `EVP_PKEY_CTX_get0_provider` worked for EdDSA. 

- Should (3) be documented? 
- Is (2) returning information about the hash **and** the asymmetric key operation, or just the hash operation? Do callers need to use both (2) and (3) to be confident that an entire hash+sign operation is being performed in the fips module? 

**The Long Form:** 

Greetings,

First, I would like to thank the OpenSSL developers for adding `EVP_PKEY_get0_provider` and `EVP_PKEY_CTX_get0_provider` The latter function works great, and avoids the extra step needed before one can use `EVP_CIPHER_get0_provider` or `EVP_MD_get0_provider`.
 
While testing the EVP_PKEY_CTX_get0_provider functionality with RSA encrypt and decrypt operations, I realized that all of my RSA and ECDSA signing and verification was covered by the "IsAlgFIPS" check, but that I had missed EdDSA.  Upon adding that call after `EVP_Digest[Sign Verify]Init`, I discovered that the call to EVP_MD_CTX_get0_md was failing with Ed25519 and Ed448, presumably because PureEdDSA operations don't actually pre-digest their message inputs.  

I found a functional workaround by digging through the documentation and evp.h --  `EVP_MD_CTX_get_pkey_ctx` -> `EVP_MD_CTX_get0_provider` but that in turn raised the question of what exactly EVP_MD_get0_provider was returning - could the hashing and asymmetric key parts of a signing or verification operation be split between two providers, and if so, would two get0_provider calls be needed to acquire the full story?

Thanks!

**Comment by t8m** [comment]
I would say: if the EVP_MD_CTX_get0_md returns NULL, you can depend on the (3) to provide sufficient assurance.
Otherwise you should check both (2) and (3).

**Comment by paulidale** [comment]
`EVP_MD_CTX_get0_md()` returns the provider for the digest.  This does not need to be the same as the PKEY provider.  I.e. you should check both to be certain.

**Comment by dskern-github** [comment]
Thank you both - I'll refactor my calling logic accordingly.
